### PR TITLE
Add testplan support

### PR DESCRIPTION
**Pull request by cnx-tcsikos**
In my company, we are using [Testplan](https://github.com/morganstanley/testplan) as the testframework, and currently there is no extension for vscode that supports it.

If we come close to the approval of the PR, I'll update the necessary docs.
Note that I'm not an experienced TS programmer, any reformatting suggestions are welcomed.

Caveats:
- Testplan doesn't report file/line information. Since the other option is to grep in the folder, I dropped this feature for performance considerations.
- There is no Testplan related configuration in 'Python' extension, so all the configuration is added to 'Python Test Explorer for VsCode'. I'm open to any alternatives.

### CODE REVIEW

File: src/testplan/testplanTestCollectionParser.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    const suites: (TestSuiteInfo | TestInfo)[] = [];
+    const parentStack: (TestSuiteInfo | TestInfo)[] = [];
+    content.split(/[\r\n]+/)
+        .map(line => line.trim())
+        .filter(line => line)
+        .map(line => line!)
+        .forEach(line => {
+            const data = line.split('::');
+            const testRank = data.length - 1;
+            if (testRank === TestObjectType.TEST)
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestCollectionParser.ts:16]: How is this working? 😅 testRank is a number right?

File: src/testplan/testplanTestRunner.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            return undefined;
+        }
+        const additionalEnvironment = await this.loadEnvironmentVariables(config);
+        this.logger.log('info', `Discovering tests using python path '${config.pythonPath()}' in ${config.getCwd()}`);
+
+        const discoveryArguments = this.getDiscoveryArguments(config.getTestplanConfiguration().testplanArguments);
+        this.logger.log('info', `Running testplan with arguments: ${discoveryArguments.join(', ')}`);
+
+        const result = await this.runTestPlan(config, additionalEnvironment, discoveryArguments).complete();
+        const tests = parseTestSuites(result.output);
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:76]: In general, there are a lot of limitations to parsing stdout. For both unittest and pytest there are some wrappers/plugins that allow having a more or less consistent output. Are there any similar options for testplan?

We can leave this as is for now, but it's definitely worth looking into such things.

File: src/testplan/testplanTestRunner.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+import { IDebugConfiguration, ITestRunner } from '../testRunner';
+import { empty } from '../utilities/collections';
+import { setDescriptionForEqualLabels } from '../utilities/tests';
+import { parseTestStates } from './testplanJunitTestStatesParser';
+import { parseTestSuites } from './testplanTestCollectionParser';
+
+// --- Testplan Exit Codes ---
+// 0: All tests were collected and passed successfully
+// 1: Tests were collected and run but some of the tests failed or none found
+const TESTPLAN_NON_ERROR_EXIT_CODES = [0, 1];
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:21]: I think this can be removed. This option was used only for pytest script wrappers.

File: src/testplan/testplanTestRunner.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    private configureCommonArgumentParser() {
+        const argumentParser = new ArgumentParser({
+            exit_on_error: false,
+        });
+        argumentParser.add_argument(
+            '--runpath',
+            { action: 'store', dest: 'runpath'});
+        argumentParser.add_argument(
+            '--timeout',
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:198]: This list of options for pytest was to filter out options that might break test discovery or execution. All other options (not from this list) will be passed to a runner (testplan in your case). So I think it's safe to remove `timeout`, `shuffle` and `verbose` option from this list.

File: src/testplan/testplanTestRunner.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+import { setDescriptionForEqualLabels } from '../utilities/tests';
+import { parseTestStates } from './testplanJunitTestStatesParser';
+import { parseTestSuites } from './testplanTestCollectionParser';
+
+// --- Testplan Exit Codes ---
+// 0: All tests were collected and passed successfully
+// 1: Tests were collected and run but some of the tests failed or none found
+const TESTPLAN_NON_ERROR_EXIT_CODES = [0, 1];
+
+const DISCOVERY_OUTPUT_PLUGIN_INFO = {
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:23]: Same here, this is useful only for pytest

**Review by kondratyev-nv — COMMENTED** [review]
Thanks for working on this! My main concern is that there is a lot of code duplication and probably some edge cases (e.g would adding `print('Hello world')` to the beginning of a test file break test discovery?). But I believe most of these concerns can be fixed later if you feel like doing it.

I see that the tests are failing, do you need any help/suggestion on fixing those?

Also, have you tried running it locally? Do you mind attaching some screenshots on how it looks?

> Testplan doesn't report file/line information.

This is ok to leave those empty, at least for now. This is mostly affecting UX - go to file/test feature won't be working.

File: src/processRunner.ts [hunk 6]
```diff
@@ -58,8 +57,6 @@ class CommandProcessExecution implements IProcessExecution {
                 if (!output) {
                     if (stdoutBuffer.length > 0) {
                         reject(new Error('Can not decode output from the process'));
-                    } else if (stderrBuffer.length > 0 && !this.commandProcess.killed) {
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply src/processRunner.ts:61]: In short, yeah.
The scenario was that if there is no testplan tests present (and testplan is enabled), it will of course try to discover the tests anyway. This is all expected.
However since testplan runs as a script, not as a module, like pytest, python will return the error
`python: can't open file 'test_plan.py': [Errno 2] No such file or directory`
to stderr, and with no stdout. Even if errno 2 is expected.
Since commandProcess.killed is already checked in the first "if", this "if" only means that "if there is no stdout, but there is stderr, reject even if error was expected"

Aaaand note to self, I removed the accepted errno 2 during the end refactor.

File: src/testplan/testplanTestCollectionParser.ts (hunk 1 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestCollectionParser.ts:16]: Yeah, this is where my missing typescript knowledge shined :D It probably can be done with maps and reduces and recursive calls.
An example result of test discovery:
`Primary
  Primary::AlphaSuite
    Primary::AlphaSuite::test_equality_passing
    Primary::AlphaSuite::test_equality_failing
    Primary::AlphaSuite::test_membership_passing
    Primary::AlphaSuite::test_membership_failing
    Primary::AlphaSuite::test_regex_passing
    Primary::AlphaSuite::test_regex_failing
Secondary
  Secondary::BetaSuite
    Secondary::BetaSuite::passing_testcase_one
    Secondary::BetaSuite::passing_testcase_two
`
Basically it iterates through the lines, testRank is the count of '::' per lines.
if testRank is
- 0, thats a App
- 1, thats a suite
- 2, thats a testcase
The order of the lines guarantee that every suite comes before its test, so what I do is store the last known app and suite and append the currently found test as its children

File: src/testplan/testplanTestRunner.ts (hunk 2 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestRunner.ts:76]: testplan provides xml output, but only for test running, not for the discovery

File: src/testplan/testplanTestRunner.ts (hunk 4 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestRunner.ts:198]: okay

File: src/testplan/testplanTestRunner.ts (hunk 3 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestRunner.ts:21]: We need the errno 2 for the described scenario above.

File: package.json [hunk 7]
```diff
@@ -117,11 +117,33 @@
...
+        "pythonTestExplorer.testplantestArgs": {
+          "type": "array",
+          "description": "Arguments passed in. Each argument is a separate item in the array.",
+          "default": [],
+          "items": {
+            "type": "string"
+          },
+          "scope": "resource"
+        },
+        "pythonTestExplorer.testplanEnabled": {
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply package.json:142]: okay

**Comment by cnx-tcsikos** [comment]
> Thanks for working on this! My main concern is that there is a lot of code duplication and probably some edge cases (e.g would adding `print('Hello world')` to the beginning of a test file break test discovery?). But I believe most of these concerns can be fixed later if you feel like doing it.
> 
> I see that the tests are failing, do you need any help/suggestion on fixing those?
> 
> Also, have you tried running it locally? Do you mind attaching some screenshots on how it looks?
> 
> > Testplan doesn't report file/line information.
> 
> This is ok to leave those empty, at least for now. This is mostly affecting UX - go to file/test feature won't be working.

I just checked, and yes, test discovery does breaks on `print("Hello world")`, that is concerning news.

There are some issues that I'm aware of:
- testplan is not installed with azure
- only install testplan on python3.6+ (any idea on this is welcomed)
- only run testplan tests on python3.6+

Locally I have only tried on windows atm, I'm hoping that when windows tests are able to run on azure, the other platforms will follow.

Here is an example img of the plugin in action (srry for the white boxes, just to be on safe side on IP)

testplan-vscode
The <build_type..> comes from parametrization annotations of testplan

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (kondratyev-nv) [reply package.json:142]: Well, if you think support both testplan and pytest at the same time for the same workspace folder is useful, then sure, let's keep it 🙂

**Comment by kondratyev-nv** [comment]
> I just checked, and yes, test discovery does breaks on print("Hello world"), that is concerning news.

Are you planning to fix this?

> There are some issues that I'm aware of:
> 
>   *  testplan is not installed with azure
>   *  only install testplan on python3.6+ (any idea on this is welcomed)
>   * only run testplan tests on python3.6+

There is no straightforward way for that as of now. If testplan can be installed with pip, you can add testplan in requirements.txt, and it seems that there is a way to filter this installation only for specific python version - https://stackoverflow.com/a/35614580/4182275 (I haven't tried it though).

Furthermore, another tricky part is to write tests that will support these conditions. I think you can add some environment variable with a python version and use this for a condition on whether to skip testplan tests (e.g. https://github.com/kondratyev-nv/vscode-python-test-adapter/blob/master/test/tests/testCancellation.test.ts#L44, but I think I saw that there is a way to do similar with just mocha).

**Comment by cnx-tcsikos** [comment]
> Are you planning to fix this?

I've created a [bug report](https://github.com/morganstanley/testplan/issues/668) for testplan. Its not that bad as I believed earlier, any `print("Hello World")` in the code happens before all the test printouts. So currently it will 'just' misinterpret the prints as test suites.
On my part what I could do as of current is to
- search the test_plan.py for any `^print\("(.*)"\)` to try and remove it from the list. This solution will only be a band-aid, as if there is more than a string literal, it won't work.
- run the testplan test with a possibly unused testname like "sfkej3adsf4ts". This will also print the unwanted `Hello World`, but will also take up some time to run the non existing test. 

> There is no straightforward way for that as of now. If testplan can be installed with pip, you can add testplan in requirements.txt, and it seems that there is a way to filter this installation only for specific python version - https://stackoverflow.com/a/35614580/4182275 (I haven't tried it though).

I made some modification to the azure-pipelines.yml to install testplan. However it does not seem to use that file, eg there is no "Install Python Dependencies" step in it at all. Is it possible that something else is used?

> Furthermore, another tricky part is to write tests that will support these conditions. I think you can add some environment variable with a python version and use this for a condition on whether to skip testplan tests (e.g. https://github.com/kondratyev-nv/vscode-python-test-adapter/blob/master/test/tests/testCancellation.test.ts#L44, but I think I saw that there is a way to do similar with just mocha).

Thanks I'll look into that.

**Comment by kondratyev-nv** [comment]
> I've created a [bug report](https://github.com/morganstanley/testplan/issues/668) for testplan. Its not that bad as I believed earlier, any `print("Hello World")` in the code happens before all the test printouts. So currently it will 'just' misinterpret the prints as test suites.
> On my part what I could do as of current is to
> 
>    * search the test_plan.py for any `^print\("(.*)"\)` to try and remove it from the list. This solution will only be a band-aid, as if there is more than a string literal, it won't work.
> 
>    * run the testplan test with a possibly unused testname like "sfkej3adsf4ts". This will also print the unwanted `Hello World`, but will also take up some time to run the non existing test.
> 

Is it possible to import testplan test with some Python API? e.g. as unittest does not have plugin support, I've created a wrapper python script that fetches all tests and outputs discovery results in easy-to-consume format - https://github.com/kondratyev-nv/vscode-python-test-adapter/blob/master/src/unittest/unittestScripts.ts#L205-L208. These docs might help - https://testplan.readthedocs.io/en/latest/multitest.html#programmatic-listing, but I'm not sure if this will cover all possible test configurations. Or maybe specify some kind of a specific pattern to distinguish regular prints from discovered tests - https://testplan.readthedocs.io/en/latest/multitest.html#command-line-listing. Let me know if this helps.

Overall, I can approve a PR if you complete testplan support even with this issue and will try fixing it in some following PRs. Let's focus on general cases first 🙂

**Comment by cnx-tcsikos** [comment]
> These docs might help - https://testplan.readthedocs.io/en/latest/multitest.html#programmatic-listing, but I'm not sure if this will cover all possible test configurations. Or maybe specify some kind of a specific pattern to distinguish regular prints from discovered tests - https://testplan.readthedocs.io/en/latest/multitest.html#command-line-listing. Let me know if this helps.

As I see it creates the some problem we want to bypass: _"Not forcing the user to do (or don't do) stuff"_. With this we could have a printout with the special `discovery_begins`-`discovery_ends` tag, but only if the user implements it, because you have to annotate your testplan app with a special lister.

> Overall, I can approve a PR if you complete testplan support even with this issue and will try fixing it in some following PRs. Let's focus on general cases first 🙂
Thanks!
I just realized that there is separately "azure pipelines" and "github action ci"

**Comment by cnx-tcsikos** [comment]
> Is it possible to import testplan test with some Python API?
I'm not sure what you are asking. There is [report package](https://testplan.readthedocs.io/en/latest/api/testplan.report.testing.html) in testplan

**Comment by kondratyev-nv** [comment]
> I'm not sure what you are asking. There is report package in testplan

The idea is doing similar to how test discovery is done for unittest (see https://github.com/kondratyev-nv/vscode-python-test-adapter/blob/master/src/unittest/unittestScripts.ts). Unittest outputs discovered tests to the stdout by default. However, there is an option to call unittest discovery programmatically which returns a set of objects that can be processed and outputted in a format that can be consumed by the extension (`defaultTestLoader.discover(start_directory, pattern=pattern)` in the script). So my question is whether it is possible to do something similar with testplan? Does it make sense? 

When adding new dependencies to package.json, please, run `npm install` to update package-lock.json. Also, prefer using `npm ci` to restore dependencies - this will ensure reproducible behavior (npm install can fetch newer versions when available, but npm ci will use version locked in package-lock.json).

I don't know why Azure pipeline is not triggered. Let's focus on GitHub Actions as those seems to be triggered fine. Both doing the same job, so I don't expect any major issues.

**Comment by cnx-tcsikos** [comment]
> > I'm not sure what you are asking. There is report package in testplan
> 
> The idea is doing similar to how test discovery is done for unittest (see https://github.com/kondratyev-nv/vscode-python-test-adapter/blob/master/src/unittest/unittestScripts.ts). Unittest outputs discovered tests to the stdout by default. However, there is an option to call unittest discovery programmatically which returns a set of objects that can be processed and outputted in a format that can be consumed by the extension (`defaultTestLoader.discover(start_directory, pattern=pattern)` in the script). So my question is whether it is possible to do something similar with testplan? Does it make sense?

Ahh okay. No, I don't think there is a way to programatically get the tests. If there was, they would have mentioned in the issue. But heck, I'll ask them.
> When adding new dependencies to package.json, please, run `npm install` to update package-lock.json. Also, prefer using `npm ci` to restore dependencies - this will ensure reproducible behavior (npm install can fetch newer versions when available, but npm ci will use version locked in package-lock.json).

Yeah, my bad :) I'm using SourceTree for git, and it showed that there was no significant changes. Usually it would mean that only EOL changes, but it was wrong.
Corrected!

> I don't know why Azure pipeline is not triggered. Let's focus on GitHub Actions as those seems to be triggered fine. Both doing the same job, so I don't expect any major issues.

Maybe because of the PR was not in "ready for review"? I did some shenanigans in the azure pipeline file blindly. A test run probably needed to check if the syntax is still right

**Comment by cnx-tcsikos** [comment]
@kondratyev-nv
One test is stuck and I can't see any logs what is it doing.
Can the PR be merged?

File: src/testplan/testplanTestRunner.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        const [, argumentsToPass] = argumentParser.parse_known_args(rawTestplanArguments);
+        return ['--info', 'pattern-full'].concat(argumentsToPass);
+    }
+
+    private getRunArguments(test: string, rawTestplanArguments: string[]): IRunArguments {
+        const argumentParser = this.configureCommonArgumentParser();
+
+        const [knownArguments, argumentsToPass] = argumentParser.parse_known_args(rawTestplanArguments);
+        return {
+            junitReportPath: (knownArguments as { xmlpath?: string }).xmlpath,
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:171]: I don't see `xmlpath` in configured arguments, so I assume it's always undefined?

File: src/testplan/testplanTestRunner.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            this.logger.log('info', 'Testplan test execution is disabled');
+            return [];
+        }
+        this.logger.log('info', `Running tests using python path '${config.pythonPath()}' in ${config.getCwd()}`);
+
+        const additionalEnvironment = await this.loadEnvironmentVariables(config);
+        const runArguments = this.getRunArguments(test, config.getTestplanConfiguration().testplanArguments);
+        const { dirName, cleanupCallback } = await this.getJunitReportPath(config.getCwd(), runArguments);
+        const testRunArguments = [
+            `--xml=${dirName}`
```
Comment (kondratyev-nv) [inline src/testplan/testplanTestRunner.ts:98]: I testplan docs I see it's possible to set `xml_dir` in tests - `@test_plan(name='SamplePlan', xml_dir='/path/to/xml-dir')`. What will happen in that case? Is `--xml` argument override this configuration?

**Review by kondratyev-nv — CHANGES_REQUESTED** [review]
Took a closer look and left some comments.

My main concern right now is about the tests. Can you please double-check that tests are executed? I assume they are always skipped because of how `isTestplanPrerequisiteMet` is implemented.

File: .azure-pipelines.yml [hunk 10]
```diff
@@ -54,6 +54,9 @@ jobs:
     displayName: 'Update pip'
   - script: python -m pip install -r requirements.txt
     displayName: 'Install requirements'
+  - script: python -m pip install https://github.com/morganstanley/testplan/archive/refs/tags/latest.zip
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply .azure-pipelines.yml:57]: I can't use the condition, if its in the requirements.txt. Testplan throws an error if its installed on a not supported python version.

File: src/testplan/testplanTestCollectionParser.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                const temp = {
+                    type: 'test' as 'test',
+                    id: data.join(':'),
+                    label: data[testRank],
+                };
+                (parentStack[testRank - 1] as TestSuiteInfo).children.push(temp);
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply src/testplan/testplanTestCollectionParser.ts:23]: can do

File: src/testplan/testplanTestRunner.ts (hunk 9 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestRunner.ts:98]: In the example it states that it overrides:
https://github.com/morganstanley/testplan/blob/0c9601b341be499a0781ae0bd935e1e9a7c69094/examples/Test%20Output/Exporters/XML/test_plan.py#L49

File: .azure-pipelines.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (kondratyev-nv) [reply .azure-pipelines.yml:57]: I assume that pip supports filtering by platform and a python version - https://stackoverflow.com/a/35614580/4182275. But I think we can keep it as is if this example is not working.

[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply .azure-pipelines.yml:57]: I didn't know that you could do this with pip, thanks!

File: package.json [hunk 12]
```diff
@@ -117,11 +118,33 @@
...
           "description": "Test framework to use for Python Test Explorer (default is null and Python extension settings are used)",
           "scope": "resource"
+        },
+        "pythonTestExplorer.testplanPath": {
+          "type": "string",
+          "default": "test_plan.py",
+          "description": "Relative path to testplan main suite.",
+          "scope": "resource"
+        },
+        "pythonTestExplorer.testplantestArgs": {
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply package.json]: whooops, corrected

File: src/testplan/testplanTestRunner.ts (hunk 8 above)
Comment (cnx-tcsikos) [reply src/testplan/testplanTestRunner.ts:172]: true, I added it to argsParser, so people can redefine its value

File: src/testplan/testplanJunitTestStatesParser.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        });
+    });
+}
+
+function parseTestResults(parserResult: any): TestEvent[] {
+    if (!parserResult) {
+        return [];
+    }
+    const testSuiteResults: ITestSuiteResult[] = parserResult.testsuites ?
+        parserResult.testsuites.testsuite : // from pytest 5.1.0, see https://github.com/pytest-dev/pytest/issues/5477
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply src/testplan/testplanJunitTestStatesParser.ts]: done

File: test/tests/utilities.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import { exec } from 'child_process';
+import { getPythonExecutable } from '../utils/testConfiguration';
+import { gte } from 'semver';
+
+
+export function isTestplanPrerequisiteMet(): boolean {
+    const command = getPythonExecutable() + ' --version';
+    exec(command, (err, stdout) => {
```
[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply test/tests/utilities.ts]: I made some changes, though its not really straightforward to me how it supposed to done correctly.
I did not wanted to put `await` in the tests bc that would require to add `es2017` to the target, so I added promise resolving thingy. It works, but looks horrible.
On the bright side, I could drop the `suite.skip` which mocha resolves as 'pending' testresult... which makes zero sense.

Let me know it you have something better solution in mind.

[the remark this answers is not in the document]
Comment (kondratyev-nv) [reply test/tests/utilities.ts]: > I did not wanted to put await in the tests bc that would require to add es2017 to the target

Do you mean for top-level await to work?

I think it's ok, at least for now. The main point is being covered with tests. My only concern is that I believe there is no guarantee that the tests would be executed even when they should, but well, I see them in CI, so let's consider this being a rare case.

[the remark this answers is not in the document]
Comment (cnx-tcsikos) [reply test/tests/utilities.ts]: > Do you mean for top-level await to work?

Yeah.

>  let's consider this being a rare case.

I managed to achieve that case, when I deleted all the other test files and left only `testplanGeneral.test.ts` in the folder. It seems that with any async test before it leaves enough time for the exec promise to finish.

**Comment by cnx-tcsikos** [comment]
Hmm I cannot replicate the errors in the CI. It seems that TestplanTestRunner::load returns undefined, which only happens when either testplan is not enabled or something banal went wrong. And testplan is enabled :D

**Comment by kondratyev-nv** [comment]
@cnx-tcsikos You can try to temporarily add `ENABLE_TEST_LOGGING=1` env to the CI test step to enable additional output.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -52,4 +52,4 @@ jobs:
     - name: Run tests
       uses: GabrielBB/xvfb-action@v1
       with:
-        run: npm test
+        run: ENABLE_TEST_LOGGING=1 npm test
```
Comment (kondratyev-nv) [inline .github/workflows/ci.yml:55]: Hm, it seems that xvfb-action is not working with additional commands... Consider using global `env:` property - https://docs.github.com/en/actions/reference/environment-variables#about-environment-variables

File: requirements.txt [hunk 16]
```diff
@@ -1,2 +1,3 @@
 pytest
 pytest-describe
+git+git://github.com/morganstanley/testplan@latest#egg=testplan
```
Comment (kondratyev-nv) [inline requirements.txt:3]: It looks like removing python restriction breaks the build for other python versions. Are there any reasons why you've removed it?

Comment (cnx-tcsikos) [reply requirements.txt:3]: just to speed up the CI result, I'll revert it when I found the issue.

**Review by kondratyev-nv — COMMENTED** [review]
Looks good! Can you please update README as a final step?

File: README.md [hunk 17]
```diff
@@ -14,20 +13,21 @@ tests with the [Test Explorer UI](https://marketplace.visualstudio.com/items?ite
...
+  * [Testplan documentation](https://testplan.readthedocs.io/en/latest/getting_started.html)
 * Open Test View sidebar
 * Run your tests using the ![Run](img/run-button.png) icon in the Test Explorer
 
 ## Features
 
 * Shows a Test Explorer in the Test view in VS Code's sidebar with all detected tests and suites and their state
 * Convenient error reporting during test discovery
-* Unittest and Pytest debugging
+* Unittest, Pytest and Testplan debugging
```
Comment (kondratyev-nv) [inline README.md:27]: Just curious, have you tested debugging?

**Review by kondratyev-nv — APPROVED** [review]

**Comment by kondratyev-nv** [comment]
It seems that cancellation tests are highly unstable. Maybe skip those for now?

File: README.md (hunk 17 above)
Comment (cnx-tcsikos) [reply README.md:27]: Great question, it is working now :)

**Comment by cnx-tcsikos** [comment]
> It seems that cancellation tests are highly unstable. Maybe skip those for now?

Now the pytestScript::should discover tests fail on macOS...

**Comment by kondratyev-nv** [comment]
Thanks for working on this! I'll try it out myself and will publish a new version in a couple of days.
